File: solutions/fable-1/fable/javascript/lcs.js
'use strict';

const data = require('fs').readFileSync(0, 'utf8');
const lines = data.split('\n').map((s) => s.replace(/\r$/, ''));
const a = lines[0] || '';
const b = lines[1] || '';

const n = a.length;
const m = b.length;

let prev = new Int32Array(m + 1);
let cur = new Int32Array(m + 1);

for (let i = 1; i <= n; i++) {
  const ai = a.charCodeAt(i - 1);
  for (let j = 1; j <= m; j++) {
    if (ai === b.charCodeAt(j - 1)) {
      cur[j] = prev[j - 1] + 1;
    } else {
      cur[j] = prev[j] > cur[j - 1] ? prev[j] : cur[j - 1];
    }
  }
  const tmp = prev;
  prev = cur;
  cur = tmp;
}

process.stdout.write(String(prev[m]) + '\n');

File: solutions/fable-1/fable/perl/lcs.pl
use strict;
use warnings;

my $a = <STDIN>;
my $b = <STDIN>;
$a = '' unless defined $a;
$b = '' unless defined $b;
$a =~ s/[\r\n]+$//;
$b =~ s/[\r\n]+$//;

my @x = split //, $a;
my @y = split //, $b;
my $n = scalar @x;
my $m = scalar @y;

my @prev = (0) x ($m + 1);
my @cur  = (0) x ($m + 1);

for my $i (1 .. $n) {
    my $ci = $x[$i - 1];
    $cur[0] = 0;
    for my $j (1 .. $m) {
        if ($ci eq $y[$j - 1]) {
            $cur[$j] = $prev[$j - 1] + 1;
        } else {
            $cur[$j] = $prev[$j] >= $cur[$j - 1] ? $prev[$j] : $cur[$j - 1];
        }
    }
    @prev = @cur;
}

print $prev[$m], "\n";
